storage/innobase/os/os0file.cc

```diff
@@ -9245,9 +9364,258 @@ Encryption::get_master_key(ulint* master_key_id,
+	} else {
+		memcpy(ptr, ENCRYPTION_KEY_MAGIC_V2, ENCRYPTION_MAGIC_SIZE);
+	}
+	ptr += ENCRYPTION_MAGIC_SIZE;
+
+	mach_write_to_4(ptr, master_key_id);
+	ptr += sizeof(ulint);
+
+	if (version == ENCRYPTION_VERSION_2) {
+		memcpy(ptr, uuid, ENCRYPTION_SERVER_UUID_LEN);
+		ptr += ENCRYPTION_SERVER_UUID_LEN;
+	}
+
+	memcpy(key_info, key, ENCRYPTION_KEY_LEN);
+
+	memcpy(key_info + ENCRYPTION_KEY_LEN, iv, ENCRYPTION_KEY_LEN);
+
+	/* Encrypt key and iv. */
+	elen = my_aes_encrypt(key_info,
+			      ENCRYPTION_KEY_LEN * 2,
+			      ptr,
+			      master_key,
+			      ENCRYPTION_KEY_LEN,
+			      my_aes_256_ecb,
+			      NULL, false);
+
+	if (elen == MY_AES_BAD_DATA) {
+		my_free(master_key);
+		return(false);
+	}
+
+	ptr += ENCRYPTION_KEY_LEN * 2;
+
+	/* Write checksum bytes. */
+	crc = ut_crc32(key_info, ENCRYPTION_KEY_LEN * 2);
+	mach_write_to_4(ptr, crc);
+
+	if (is_boot) {
+		ut_free(master_key);
+	} else {
+		my_free(master_key);
+	}
+
+	return(true);
+}
+
+/** Decoding the encryption info
+from the first page of a tablespace.
+@param[in,out]	key		key
+@param[in,out]	iv		iv
+@param[in]	encryption_info	encrytion info.
+@return true if success */
+bool
+Encryption::decode_encryption_info(byte*	key,
+				   byte*	iv,
+				   byte*	encryption_info)
+{
+	byte*			ptr;
+	ulint			m_key_id;
+	byte*			master_key = NULL;
+	lint			elen;
+	byte			key_info[ENCRYPTION_KEY_LEN * 2];
+	ulint			crc1;
+	ulint			crc2;
+	char			srv_uuid[ENCRYPTION_SERVER_UUID_LEN + 1];
+	Version			version;
+#ifdef	UNIV_ENCRYPT_DEBUG
+	const byte*		data;
+	ulint			i;
+#endif
+
+	ptr = encryption_info;
+
+	/* For compatibility with 5.7.11, we need to handle the
+	encryption information which created in this old version. */
+	if (memcmp(ptr, ENCRYPTION_KEY_MAGIC_V1,
+		     ENCRYPTION_MAGIC_SIZE) == 0) {
+		version = ENCRYPTION_VERSION_1;
+	} else {
+		version = ENCRYPTION_VERSION_2;
+	}
+
+	/* Check magic. */
+	if (version == ENCRYPTION_VERSION_2
+	    && memcmp(ptr, ENCRYPTION_KEY_MAGIC_V2, ENCRYPTION_MAGIC_SIZE) != 0) {
+		/* We ignore report error for recovery,
+		since the encryption info maybe hasn't writen
+		into datafile when the table is newly created. */
+		if (!recv_recovery_is_on()) {
+			return(false);
+		} else {
+			return(true);
+		}
+	}
+
+	ptr += ENCRYPTION_MAGIC_SIZE;
+
+	/* Get master key id. */
+	m_key_id = mach_read_from_4(ptr);
+	ptr += sizeof(ulint);
+
+	/* Get server uuid. */
+	if (version == ENCRYPTION_VERSION_2) {
+		memset(srv_uuid, 0, ENCRYPTION_SERVER_UUID_LEN + 1);
+		memcpy(srv_uuid, ptr, ENCRYPTION_SERVER_UUID_LEN);
+		ptr += ENCRYPTION_SERVER_UUID_LEN;
+	}
+
+	/* Get master key by key id. */
+	memset(key_info, 0, ENCRYPTION_KEY_LEN * 2);
+	if (version == ENCRYPTION_VERSION_1) {
+		get_master_key(m_key_id, NULL, &master_key);
+	} else {
+		if (m_key_id == 0) {
+			/* When m_key_id is 0, which means it's the
+			default master key for bootstrap. */
+			master_key = static_cast<byte*>(ut_zalloc_nokey(
```

Comment:
Consider using my_ allocation

Reply:
So what is your opinion about considering using my_ allocation?
(by now only change something if you  have to re-push/re-test otherwise)